chain/client/src/client_actor.rs

```diff
@@ -1708,65 +1700,110 @@ impl ClientActor {
-                            {
-                                let id = peer_info.peer_info.id.clone();
-
-                                if let Ok(header) =
-                                    self.client.chain.get_block_header(&state_sync_status.sync_hash)
-                                {
-                                    for hash in
-                                        vec![*header.prev_hash(), *header.hash()].into_iter()
-                                    {
-                                        self.client.request_block(hash, id.clone());
-                                    }
+                    let now = StaticClock::utc();
+
+                    // FIXME: it checks if the block exists.. but I have no idea why..
+                    // seems that we don't really use this block in case of catchup - we use it only for state sync.
+                    // Seems it is related to some bug with block getting orphaned after state sync? but not sure.
+                    let (request_block, have_block) =
+                        unwrap_and_report!(self.sync_block_status(&prev_hash, now));
+
+                    if request_block {
+                        self.client.state_sync.last_time_block_requested = Some(now);
+                        if let Some(peer_info) =
+                            self.network_info.highest_height_peers.choose(&mut thread_rng())
+                        {
+                            let id = peer_info.peer_info.id.clone();
+
+                            if let Ok(header) = self.client.chain.get_block_header(&sync_hash) {
+                                for hash in vec![*header.prev_hash(), *header.hash()].into_iter() {
+                                    self.client.request_block(hash, id.clone());
                                 }
                             }
                         }
-                        StateSyncResult::Completed => {
-                            info!(target: "sync", "State sync: all shards are done");
-
-                            let mut block_processing_artifacts = BlockProcessingArtifact::default();
-
-                            unwrap_and_report!(self.client.chain.reset_heads_post_state_sync(
-                                &me,
-                                state_sync_status.sync_hash,
-                                &mut block_processing_artifacts,
-                                self.get_apply_chunks_done_callback(),
-                            ));
-
-                            self.client
-                                .process_block_processing_artifact(block_processing_artifacts);
-
-                            self.client.sync_status = SyncStatus::BodySync {
-                                start_height: 0,
-                                current_height: 0,
-                                highest_height: 0,
-                            };
+                    }
+
+                    if have_block {
```

Comment:
doesn't this `if have_block { }` make this not just a no-op refactor? In the previous logic, if we don't have the block, we still proceed to run `StateSync::sync_shards_status()`, and we'll send `StateRequestHeader`s in the meantime. Now after this change we'll wait until we have the whole `sync_hash` block until we do any of that. That's not an intended change is it? Seems like we might as well keep the old behavior and just move on with the rest of state sync until we get the block

Reply:
Good catch! Thanks for pointing it out! We should go through state sync but not proceed after unless we have the sync block.